runtime/parachains/src/paras_inherent/mod.rs

```diff
@@ -294,218 +269,258 @@ pub mod pallet {
+		let mut rng = rand_chacha::ChaChaRng::from_seed(entropy.into());
 
-			// .. and prepare a helper closure.
-			let dispute_set_validity_check = move |set| {
-				T::DisputesHandler::filter_dispute_data(
-					set,
-					post_conclusion_acceptance_period,
-					verify_dispute_sigs,
-				)
-			};
+		// Filter out duplicates and continue.
+		if let Err(()) = T::DisputesHandler::deduplicate_and_sort_dispute_data(&mut disputes) {
+			log::debug!(target: LOG_TARGET, "Found duplicate statement sets, retaining the first");
+		}
 
-			// In case of an overweight block, consume up to the entire block weight
-			// in disputes, since we will never process anything else, but invalidate
-			// the block. It's still reasonable to protect against a massive amount of disputes.
-			if candidates_weight
-				.saturating_add(bitfields_weight)
-				.saturating_add(disputes_weight)
-				.any_gt(max_block_weight)
-			{
-				log::warn!("Overweight para inherent data reached the runtime {:?}", parent_hash);
-				backed_candidates.clear();
-				candidates_weight = Weight::zero();
-				signed_bitfields.clear();
-				bitfields_weight = Weight::zero();
-			}
+		let config = <configuration::Pallet<T>>::config();
+		let post_conclusion_acceptance_period = config.dispute_post_conclusion_acceptance_period;
 
-			let entropy = compute_entropy::<T>(parent_hash);
-			let mut rng = rand_chacha::ChaChaRng::from_seed(entropy.into());
+		let dispute_statement_set_valid = move |set: DisputeStatementSet| {
+			T::DisputesHandler::filter_dispute_data(set, post_conclusion_acceptance_period)
+		};
 
-			let (checked_disputes, checked_disputes_weight) = limit_and_sanitize_disputes::<T, _>(
+		// Limit the disputes first, since the following statements depend on the votes include here.
+		let (checked_disputes_sets, checked_disputes_sets_consumed_weight) =
+			limit_and_sanitize_disputes::<T, _>(
 				disputes,
-				&dispute_set_validity_check,
+				dispute_statement_set_valid,
 				max_block_weight,
 				&mut rng,
 			);
-			(
-				checked_disputes,
-				checked_disputes_weight
-					.saturating_add(candidates_weight)
-					.saturating_add(bitfields_weight),
-			)
-		};
 
-		let expected_bits = <scheduler::Pallet<T>>::availability_cores().len();
+		// Assure the maximum block weight is adhered, by limiting bitfields and backed
+		// candidates. Dispute statement sets were already limited before.
+		let actual_weight = apply_weight_limit::<T>(
+			&mut backed_candidates,
+			&mut bitfields,
+			max_block_weight.saturating_sub(checked_disputes_sets_consumed_weight),
+			&mut rng,
+		);
 
-		// Handle disputes logic.
-		let disputed_bitfield = {
-			let new_current_dispute_sets: Vec<_> = checked_disputes
-				.iter()
-				.map(AsRef::as_ref)
-				.filter(|s| s.session == current_session)
-				.map(|s| (s.session, s.candidate_hash))
-				.collect();
+		if actual_weight.any_gt(max_block_weight) {
+			log::warn!(target: LOG_TARGET, "Post weight limiting weight is still too large.");
```

Comment:
I don't get how is this possible, assuming `apply_weight_limit` limits backed candidates and bitfields correctly. The only was I could see this happening is checked_disputes_sets_consumed_weight > `max_block_weight`, but this implies that dispute filtering process is broken as well.

Reply:
I just kept the check, indeed it should not be possible - nevertheless a good sanity check. I fixed it to use the full weight though and not just the ones of bitfields and candidates.